umpost/conversion_driver_esm1p5.py

```diff
@@ -127,28 +128,88 @@ def convert_fields_file_list(fields_file_paths, nc_write_dir):
 
     Returns
     -------
-    None
+    succeeded: list of tuples of (input, output) filepaths for successful
+    conversions.
+    failed: list of tuples of form (input path, exception) for files which failed 
+    to convert due to an allowed exception. 
     """
+    succeeded = []
+    failed = []
 
-    for fields_file_path in fields_file_paths:
+    fields_file_paths = [Path(p) for p in fields_file_paths]
 
-        fields_file_path = Path(fields_file_path)
+    for fields_file_path in fields_file_paths:
 
         nc_write_path = get_nc_write_path(fields_file_path, nc_write_dir)
 
-        print("Converting file " + fields_file_path.name)
-
         try:
             um2netcdf4.process(fields_file_path, nc_write_path, ARG_VALS)
+            succeeded.append((fields_file_path, nc_write_path))
 
         except Exception as exc:
-            # Not ideal here - um2netcdf4 raises generic exception when missing coordinates
+            # TODO: Refactor once um2nc has specific exceptions
             if exc.args[0] in ALLOWED_UM2NC_EXCEPTION_MESSAGES.values():
-                warnings.warn("Unable to convert file: " +
-                              fields_file_path.name)
+                failed.append((fields_file_path, exc))
             else:
+                # raise any unexpected errors
                 raise
 
+    return succeeded, failed
+
+
+def format_successes(succeeded):
+    """
+    Format reports of successful conversions to be shared with user.
+
+    Parameters
+    ---------- 
+    succeeded: list of tuples of (input, output) filepaths for successful
+    conversions.
+
+    Returns
+    -------
+    success_reports: list of formatted reports of successful conversions.
+    """
+    success_reports = []
+    
+    for input_path, output_path in succeeded:
+        success_reports.append(
+            f"Successfully converted {output_path}"
+        )
+
+    return success_reports
+
+
+def format_failures(failed, quiet):
+    """
+    Format reports of conversions which failed with permitted exceptions.
+
+    Parameters
+    ----------
+    failed: list of tuples of form (filepath, exception) for files which failed 
+    to convert due to an allowed exception. 
+    quiet: boolean. Report only final exception type and message rather than 
+    full stack trace when true.
+
+    Returns
+    -------
+    failure_reports: list of formatted reports of failed conversions.
+    """
+    failure_reports = []
+    for fields_file_path, exception in failed:
+        report_base = f"Failed to convert {fields_file_path}. "
+        if quiet:
+            report = report_base + "Final reported error: \n" + repr(exception)
+        else:
+            formatted_traceback = "".join(
+                traceback.format_exception(exception)
+            )
+            report = report_base + "Stack trace: \n" + formatted_traceback
+
+        failure_reports.append(report)
+
+    return failure_reports
```

Comment:
I had similar thoughts about converting this to a generator & other mods:
1. The `if quiet` check is called on every loop iteration. This check can be "promoted" to the first/one off bool check, with the branches calling the loops.
2. It's less of a concern in this project, but concatenating strings is inefficient. One concise fix is to tweak the report format slightly: `yield report_base`, then `yield f"Final reported error ... {something}\n"`. This avoids the `str + str` & the caller simply prints some lines. This should to work the the tracebacks too.
3. Converting to a generator reduces memory use as report funcs emit the report gradually instead of creating it all at once.

Reply:
Good idea. I've moved the `if quiet` check outside of the loop. 

For the string concatenation, is it ok to push those changes down the track? I think one challenge with the failure reporting, is that the final failure report is given to the user as a warning:
```python
    for failure_message in format_failures(failed, args.quiet):
        warnings.warn(failure_message)
```
and so it might be difficult to avoid combining the different parts of the reports together into a single string.

Reply:
In https://github.com/ACCESS-NRI/um2nc-standalone/commit/440cf29f2de1324f93a9d4ecd24b304, it looks like you're yielding 1 report per failure, which should fit into a warning.

If the failure report was generating individual lines, something like `print(failure_msg, file=sys.stderr)` should work. This touches `sys.stderr`, which isn't quite as obvious as using the `warnings` module.